src/api/c/sparse.cpp

```diff
@@ -0,0 +1,395 @@
+        af_sparse_t sparse;
+        sparse.storage = storage;
+        sparse.nRows = nRows;
+        sparse.nCols = nCols;
+        sparse.nNZ = nNZ;
+
+        AF_CHECK(af_retain_array(&sparse.rowIdx, rowIdx));
+        AF_CHECK(af_retain_array(&sparse.colIdx, colIdx));
+        AF_CHECK(af_retain_array(&sparse.values, values));
+
+        *out = getSparseHandle(sparse);
+    } CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_create_sparse_array_from_host(
+                 af_sparse_array *out,
+                 const dim_t nRows, const dim_t nCols, const dim_t nNZ,
+                 const void * const values,
+                 const int * const rowIdx, const int * const colIdx,
+                 const af_dtype type, const af_sparse_storage storage)
+{
+    try {
+        // Checks:
+        // rowIdx and colIdx arrays are of s32 type
+        // values is of floating point type
+        // if COO, rowIdx, colIdx and values should have same dims
+        // if CRS, colIdx and values should have same dims, rowIdx.dims = nRows
+        // if CRC, rowIdx and values should have same dims, colIdx.dims = nCols
+        // storage is within acceptable range
+        // type is floating type
+        if(!(storage == AF_SPARSE_COO
+          || storage == AF_SPARSE_CSR
+          || storage == AF_SPARSE_CSC)) {
+            AF_ERROR("Storage type is out of range/unsupported", AF_ERR_ARG);
+        }
+
+        TYPE_ASSERT(type == f32 || type == f64
+                 || type == c32 || type == c64);
+
+        af_sparse_t sparse;
+        sparse.storage = storage;
+        sparse.nRows = nRows;
+        sparse.nCols = nCols;
+        sparse.nNZ = nNZ;
+
+        AF_CHECK(af_create_array(&sparse.values, values, 1, &nNZ, type));
+
+        if(storage == AF_SPARSE_COO) {
+            AF_CHECK(af_create_array(&sparse.rowIdx, rowIdx, 1, &nNZ, s32));
+            AF_CHECK(af_create_array(&sparse.colIdx, colIdx, 1, &nNZ, s32));
+        } else if(storage == AF_SPARSE_CSR) {
+            AF_CHECK(af_create_array(&sparse.rowIdx, rowIdx, 1, &nRows, s32));
+            AF_CHECK(af_create_array(&sparse.colIdx, colIdx, 1, &nNZ, s32));
+        } else if(storage == AF_SPARSE_CSC) {
+            AF_CHECK(af_create_array(&sparse.rowIdx, rowIdx, 1, &nNZ, s32));
+            AF_CHECK(af_create_array(&sparse.colIdx, colIdx, 1, &nCols, s32));
+        }
+
+        *out = getSparseHandle(sparse);
+    } CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+template<typename T, af_sparse_storage storage>
+void create_sparse_array_from_dense2(af_sparse_t *out, const af_array in_, const af_array nonZeroIdx_)
+{
+    Array<int> nonZeroIdx = castArray<int>(nonZeroIdx_);
+
+    const Array<T> in = getArray<T>(in_);
+
+    dim_t nNZ = nonZeroIdx.elements();
+    Array<int> constNNZ = createValueArray<int>(dim4(nNZ), nNZ);
+
+    Array<int> rowIdx = *initArray<int>();
+    Array<int> colIdx = *initArray<int>();
+    Array<T>   values = *initArray<T>();
+
+    if(storage == AF_SPARSE_COO) {
```

Comment:
storage does not need to be a template for this function. Merge `create_sparse_array_from_dense{1,2}`